Don't export libpq.so symbols in the postgres binary.

lmzzzzz1 opened the pull request:
Due to Greenplum's MPP architecture, the postgres process needs to connect to
databases, some symbols from libpq.so are compiled into the postgres binary as
well. However, some symbols are not safe to be exported both in the postgres
binary and the libpq.so library. E.g., some of the functions are using
palloc()/pfree() to allocate/free memory in backend codes while using
malloc()/free() in frontend codes. See issue #16219. This patch resolves the
issue by hiding frontend symbols in the postgres binary, so that frontend
symbols referenced in extensions will be resolved correctly.

---- Original description ----

In fe-auth-scram.c, we use the macro FRONTED to isolate the front-end and back-end codes to prevent freeing the memory that has been pfreed again

## Here are some reminders before you submit the pull request
- [x] Add tests for the change
- [x] Document changes
- [x] Communicate in the mailing list if needed
- [x] Pass `make installcheck`
- [x] Review a PR in return to support the community

hyongtao-db reviewed (approved):
Looks good!

zhaorui20005 commented:
After checking the structure of the `fe_scram_state`, only the `password` has the chance to be alloced by palloc, and this fix is simple. I think this PR looks good. I will start a full pipeline for this branch.

Neustradamus commented:
@lmzzzzz1: Good job!

Linked to:
- https://github.com/scram-sasl/info/issues/1

zhaorui20005 commented:
```diff -I HINT: -I CONTEXT: -I GP_IGNORE: -U3 /tmp/build/e18b2f02/gpdb_src/contrib/postgres_fdw/expected/gp_postgres_fdw.out /tmp/build/e18b2f02/gpdb_src/contrib/postgres_fdw/results/gp_postgres_fdw.out
--- /tmp/build/e18b2f02/gpdb_src/contrib/postgres_fdw/expected/gp_postgres_fdw.out	2023-08-18 05:02:36.522316222 +0000
+++ /tmp/build/e18b2f02/gpdb_src/contrib/postgres_fdw/results/gp_postgres_fdw.out	2023-08-18 05:02:36.598324170 +0000
@@ -2119,6 +2119,7 @@
 
 \c postgres
 create user u1 password '123456';
+ERROR:  role "u1" already exists
 create database database_16219;
 \c database_16219
 create extension postgres_fdw;
@@ -2140,11 +2141,8 @@
 CREATE FOREIGN TABLE f_t1(a int, b int) 
 	server database_16219 options(schema_name 'public', table_name 't1');
 select count(*) from f_t1;
- count 
--------
-    10
-(1 row)
-
+DETAIL:  FATAL:  password authentication failed for user "u1"
+ERROR:  could not connect to server "database_16219"
 DO $d$
     BEGIN
         EXECUTE $$ALTER SERVER database_16219
@@ -2152,11 +2150,8 @@
     END;
 $d$;
 select count(*) from f_t1;
- count 
--------
-    10
-(1 row)
-
+DETAIL:  FATAL:  password authentication failed for user "u1"
+ERROR:  could not connect to server "database_16219"
 \c postgres
 drop database database_16219;
 drop user u1;


real	133m36.712s
user	16m2.069s
sys	7m5.748s
```
Hi, @lmzzzzz1 the testing cases failed, the reason maybe the u1 has been created by previous cases. I think you need to create an uniq user name.

lmzzzzz1 commented:
OK, I updated the test case.

zhaorui20005 reviewed (approved):
Thanks for your contribution.

higuoxing commented:
Hi, 

Due to Greenplum's MPP architecture, the coordinator itself needs to connect to databases on segments and some of the frontend codes get compiled into the `postgres` binary (just as you mentioned in https://github.com/greenplum-db/gpdb/issues/16219#issuecomment-1681866571). The problem is that these symbols are defined both in `postgres` binary and `libpq.so`  and their visibility are 'GLOBAL'. E.g.,

```
➜ objdump -t bin/postgres | grep PQconnectdbParams
00000000006d6b80 g     F .text  0000000000000024              PQconnectdbParams

➜ objdump -t /tmp/gpdb7/lib/libpq.so | grep PQconnectdbParams
000000000000b520 g     F .text  0000000000000024              PQconnectdbParams
```

I'm not sure if there will be similar problems when the coordinator connects to segments. But for extensions, we can teach them to find front-end symbols from `libpq.so` rather than `postgres`. To achieve this, one of the solution is making these duplicated symbols (defined in https://github.com/greenplum-db/gpdb/blob/main/src/interfaces/libpq/exports.txt)
local in `postgres` and global in `libpq.so`. We can use linker script to do this job. I have a patch for you to play with. What do you think of it? The linker script itself can be generated from https://github.com/greenplum-db/gpdb/blob/main/src/interfaces/libpq/exports.txt when we build gpdb.

With my patch, the visibility of these symbols is changed to,

```
➜ objdump -t bin/postgres | grep PQconnectdbParams
00000000006d5b80 l     F .text  0000000000000024              PQconnectdbParams
➜ objdump -t lib/libpq.so | grep PQconnectdbParams
000000000000b520 g     F .text  0000000000000024              PQconnectdbParams
```

The patch:

```diff
From 9a652404d1f9c5286b0d1b8a8a9a796d43a11b61 Mon Sep 17 00:00:00 2001
From: Xing Guo <higuoxing@gmail.com>
Date: Sat, 19 Aug 2023 00:20:41 +0800
Subject: [PATCH v1] hack

---
 gpcontrib/gp_exttable_fdw/Makefile |   1 +
 gpcontrib/gp_inject_fault/Makefile |   4 +-
 src/backend/Makefile               |   2 +-
 src/backend/exported_symbol.list   | 180 +++++++++++++++++++++++++++++
 4 files changed, 185 insertions(+), 2 deletions(-)
 create mode 100644 src/backend/exported_symbol.list

diff --git a/gpcontrib/gp_exttable_fdw/Makefile b/gpcontrib/gp_exttable_fdw/Makefile
index d7e171d53b..c6d2ba125a 100644
--- a/gpcontrib/gp_exttable_fdw/Makefile
+++ b/gpcontrib/gp_exttable_fdw/Makefile
@@ -1,6 +1,7 @@
 MODULE_big = gp_exttable_fdw
 OBJS = gp_exttable_fdw.o extaccess.o option.o
 
+SHLIB_LINK_INTERNAL = $(libpq)
 #SHLIB_LINK += -lcurl
 
 EXTENSION = gp_exttable_fdw
diff --git a/gpcontrib/gp_inject_fault/Makefile b/gpcontrib/gp_inject_fault/Makefile
index b7991ac9f5..0539a262d0 100644
--- a/gpcontrib/gp_inject_fault/Makefile
+++ b/gpcontrib/gp_inject_fault/Makefile
@@ -1,4 +1,5 @@
-MODULES = gp_inject_fault
+MODULE_big = gp_inject_fault
+OBJS = gp_inject_fault.o
 
 EXTENSION = gp_inject_fault
 DATA = gp_inject_fault--1.0.sql
@@ -6,6 +7,7 @@ DATA = gp_inject_fault--1.0.sql
 REGRESS = inject_fault_test
 
 PG_CPPFLAGS = -I$(libpq_srcdir)
+SHLIB_LINK_INTERNAL = $(libpq)
 
 ifdef USE_PGXS
 PG_CONFIG = pg_config
diff --git a/src/backend/Makefile b/src/backend/Makefile
index b4efac73d4..b7a31d6b07 100644
--- a/src/backend/Makefile
+++ b/src/backend/Makefile
@@ -72,7 +72,7 @@ ifneq ($(PORTNAME), win32)
 ifneq ($(PORTNAME), aix)
 
 postgres: $(OBJS)
-	$(CXX) $(CXXFLAGS) $(call expand_subsys,$^) $(LDFLAGS) $(LDFLAGS_EX) $(export_dynamic) $(LIBS) -o $@
+	$(CXX) $(CXXFLAGS) $(call expand_subsys,$^) $(LDFLAGS) $(LDFLAGS_EX) $(export_dynamic) $(LIBS) -Wl,--version-script=exported_symbol.list -o $@
 
 endif
 endif
diff --git a/src/backend/exported_symbol.list b/src/backend/exported_symbol.list
new file mode 100644
index 0000000000..52aea7d88f
--- /dev/null
+++ b/src/backend/exported_symbol.list
@@ -0,0 +1,180 @@
+{ global: *;
+
+ local:
+PQconnectdb;
+PQsetdbLogin;
+PQconndefaults;
+PQfinish;
+PQreset;
+PQrequestCancel;
+PQdb;
+PQuser;
+PQpass;
+PQhost;
+PQport;
+PQtty;
+PQoptions;
+PQstatus;
+PQerrorMessage;
+PQsocket;
+PQbackendPID;
+PQtrace;
+PQuntrace;
+PQsetNoticeProcessor;
+PQexec;
+PQnotifies;
+PQsendQuery;
+PQgetResult;
+PQisBusy;
+PQconsumeInput;
+PQgetline;
+PQputline;
+PQgetlineAsync;
+PQputnbytes;
+PQendcopy;
+PQfn;
+PQresultStatus;
+PQntuples;
+PQnfields;
+PQbinaryTuples;
+PQfname;
+PQfnumber;
+PQftype;
+PQfsize;
+PQfmod;
+PQcmdStatus;
+PQoidStatus;
+PQcmdTuples;
+PQgetvalue;
+PQgetlength;
+PQgetisnull;
+PQclear;
+PQmakeEmptyPGresult;
+PQprint;
+PQdisplayTuples;
+PQprintTuples;
+lo_open;
+lo_close;
+lo_read;
+lo_write;
+lo_lseek;
+lo_creat;
+lo_tell;
+lo_unlink;
+lo_import;
+lo_export;
+pgresStatus;
+PQmblen;
+PQresultErrorMessage;
+PQresStatus;
+termPQExpBuffer;
+appendPQExpBufferChar;
+initPQExpBuffer;
+resetPQExpBuffer;
+PQoidValue;
+PQclientEncoding;
+PQenv2encoding;
+appendBinaryPQExpBuffer;
+appendPQExpBufferStr;
+destroyPQExpBuffer;
+createPQExpBuffer;
+PQconninfoFree;
+PQconnectPoll;
+PQconnectStart;
+PQflush;
+PQisnonblocking;
+PQresetPoll;
+PQresetStart;
+PQsetClientEncoding;
+PQsetnonblocking;
+PQfreeNotify;
+PQescapeString;
+PQescapeBytea;
+printfPQExpBuffer;
+appendPQExpBuffer;
+pg_encoding_to_char;
+pg_utf_mblen;
+PQunescapeBytea;
+PQfreemem;
+PQtransactionStatus;
+PQparameterStatus;
+PQprotocolVersion;
+PQsetErrorVerbosity;
+PQsetNoticeReceiver;
+PQexecParams;
+PQsendQueryParams;
+PQputCopyData;
+PQputCopyEnd;
+PQgetCopyData;
+PQresultErrorField;
+PQftable;
+PQftablecol;
+PQfformat;
+PQexecPrepared;
+PQsendQueryPrepared;
+PQdsplen;
+PQserverVersion;
+PQgetssl;
+pg_char_to_encoding;
+pg_valid_server_encoding;
+pqsignal;
+PQprepare;
+PQsendPrepare;
+PQgetCancel;
+PQfreeCancel;
+PQcancel;
+lo_create;
+PQinitSSL;
+PQregisterThreadLock;
+PQescapeStringConn;
+PQescapeByteaConn;
+PQencryptPassword;
+PQisthreadsafe;
+enlargePQExpBuffer;
+PQnparams;
+PQparamtype;
+PQdescribePrepared;
+PQdescribePortal;
+PQsendDescribePrepared;
+PQsendDescribePortal;
+lo_truncate;
+PQconnectionUsedPassword;
+pg_valid_server_encoding_id;
+PQconnectionNeedsPassword;
+lo_import_with_oid;
+PQcopyResult;
+PQsetResultAttrs;
+PQsetvalue;
+PQresultAlloc;
+PQregisterEventProc;
+PQinstanceData;
+PQsetInstanceData;
+PQresultInstanceData;
+PQresultSetInstanceData;
+PQfireResultCreateEvents;
+PQconninfoParse;
+PQinitOpenSSL;
+PQescapeLiteral;
+PQescapeIdentifier;
+PQconnectdbParams;
+PQconnectStartParams;
+PQping;
+PQpingParams;
+PQlibVersion;
+PQsetSingleRowMode;
+lo_lseek64;
+lo_tell64;
+lo_truncate64;
+PQconninfo;
+PQsslInUse;
+PQsslStruct;
+PQsslAttributeNames;
+PQsslAttribute;
+PQsetErrorContextVisibility;
+PQresultVerboseErrorMessage;
+PQencryptPasswordConn;
+PQresultMemorySize;
+PQhostaddr;
+PQgssEncInUse;
+PQgetgssctx;
+};
-- 
2.41.0

```

higuoxing reviewed (commented):
> Hi, @higuoxing I think this solution is better and can solve the problem fundamentally. In addition, I found that the modification of the linker is very similar to the upstream discussion. https://www.postgresql.org/message-id/153626613985.23143.4743626885618266803%40wrigleys.postgresql.org

Thanks for the information.

> The only problem is to maintain both export.txt and exported_symbol.list files.

Yeah, it's a quick patch and `exported_symbol.list` can be generated from `export.txt` . The linker script for the `postgres` binary can be generated in a similar way to building linker script for `libpq.so`. Except that we should hide symbols from `postgres` that are exported from `libpq.so` (Final exported symbols from postgres = Initial exported symbols from postgres - Exported symbols from libpq.so).

Below is the command for generating linker script from `export.txt` for libpq.so:

https://github.com/greenplum-db/gpdb/blob/c5abb1f3e6c20f5fa372c2c807dfd22b350985f2/src/Makefile.shlib#L234

----------------------------

Would you like to continue the work since you've already had great test case and explored this topic in depth?

lmzzzzz1 commented:
> > Hi, @higuoxing I think this solution is better and can solve the problem fundamentally. In addition, I found that the modification of the linker is very similar to the upstream discussion. https://www.postgresql.org/message-id/153626613985.23143.4743626885618266803%40wrigleys.postgresql.org
> 
> Thanks for the information.
> 
> > The only problem is to maintain both export.txt and exported_symbol.list files.
> 
> Yeah, it's a quick patch and `exported_symbol.list` can be generated from `export.txt` . The linker script for the `postgres` binary can be generated in a similar way to building linker script for `libpq.so`. Except that we should hide symbols from `postgres` that are exported from `libpq.so` (Final exported symbols from postgres = Initial exported symbols from postgres - Exported symbols from libpq.so).
> 
> Below is the command for generating linker script from `export.txt` for libpq.so:
> 
> https://github.com/greenplum-db/gpdb/blob/c5abb1f3e6c20f5fa372c2c807dfd22b350985f2/src/Makefile.shlib#L234
> 
> Would you like to continue the work since you've already had great test case and explored this topic in depth?

OK, I will try this way.

zhaorui20005 commented:
Is it ok for us to push commits to your branch @lmzzzzz1 ? If it is ok, I think @higuoxing could push your commits to this PR, since you have already tested with your local env.

lmzzzzz1 commented:
> Is it ok for us to push commits to your branch @lmzzzzz1 ? If it is ok, I think @higuoxing could push your commits to this PR, since you have already tested with your local env.

Yes, I modified the backend/Makefile according to @higuoxing , generated the export.list needed by GPDB, and updated the PR, you can directly modify and verify it on my PR.

zhaorui20005 commented:
Here is the pull pipeline address: https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/gpdb-dev-yjw/
There are many testing cases failed, I just put some of them for your information:
```
ERROR:  could not load library "/usr/local/greenplum-db-devel/lib/postgresql/plperl.so": /usr/local/greenplum-db-devel/lib/postgresql/plperl.so: undefined symbol: pqsignal
command failed: "/usr/local/greenplum-db-devel/bin/psql" -X -c "CREATE EXTENSION IF NOT EXISTS \"plperl\"" "pl_regression"
+ERROR:  could not load library "/usr/local/greenplum-db-devel/lib/postgresql/orafce.so": /usr/local/greenplum-db-devel/lib/postgresql/orafce.so: undefined symbol: pg_char_to_encoding
ERROR:  could not load library "/tmp/build/e18b2f02/gpdb_src/src/test/regress/regress.so": /tmp/build/e18b2f02/gpdb_src/src/test/regress/regress.so: undefined symbol: PQresultStatus (dfmgr.c:249)
ERROR:  could not load library "/usr/local/greenplum-db-devel/lib/postgresql/utf8_and_iso8859_1.so": /usr/local/greenplum-db-devel/lib/postgresql/utf8_and_iso8859_1.so: undefined symbol: pg_utf_mblen
```

higuoxing commented:
> Here is the pull pipeline address: https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/gpdb-dev-yjw/
There are many testing cases failed, I just put some of them for your information:

```
ERROR:  could not load library "/usr/local/greenplum-db-devel/lib/postgresql/plperl.so": /usr/local/greenplum-db-devel/lib/postgresql/plperl.so: undefined symbol: pqsignal
command failed: "/usr/local/greenplum-db-devel/bin/psql" -X -c "CREATE EXTENSION IF NOT EXISTS \"plperl\"" "pl_regression"
+ERROR:  could not load library "/usr/local/greenplum-db-devel/lib/postgresql/orafce.so": /usr/local/greenplum-db-devel/lib/postgresql/orafce.so: undefined symbol: pg_char_to_encoding
ERROR:  could not load library "/tmp/build/e18b2f02/gpdb_src/src/test/regress/regress.so": /tmp/build/e18b2f02/gpdb_src/src/test/regress/regress.so: undefined symbol: PQresultStatus (dfmgr.c:249)
ERROR:  could not load library "/usr/local/greenplum-db-devel/lib/postgresql/utf8_and_iso8859_1.so": /usr/local/greenplum-db-devel/lib/postgresql/utf8_and_iso8859_1.so: undefined symbol: pg_utf_mblen
```

Ah, seems that some symbols are defined both in the postgres binary and libpq.so. We should setup some kinds of exception-rules for them in the linker script.

higuoxing commented:
I'm running out of time today. Will review changes tomorrow.

higuoxing commented:
Hi, @lmzzzzz1 I added an additional commit to fix the failure mentioned above. Could you please take a look? I'm not a Makefile expert so I may make mistakes :-P

Symbols like `pqsignal` and those whose name starts with `pg_` should be exported both in `libpq.so` and the postgres binary. Things left us to do:

- Can we simplify the test? The current test case is failing on our CI system.

```diff
======================================================================
DIFF FILE: ./contrib/postgres_fdw/regression.diffs
----------------------------------------------------------------------

diff -I HINT: -I CONTEXT: -I GP_IGNORE: -U3 /home/gpadmin/gpdb_src/contrib/postgres_fdw/expected/gp_postgres_fdw_optimizer.out /home/gpadmin/gpdb_src/contrib/postgres_fdw/results/gp_postgres_fdw.out
--- /home/gpadmin/gpdb_src/contrib/postgres_fdw/expected/gp_postgres_fdw_optimizer.out	2023-08-23 05:22:54.418706466 +0000
+++ /home/gpadmin/gpdb_src/contrib/postgres_fdw/results/gp_postgres_fdw.out	2023-08-23 05:22:54.482712330 +0000
@@ -2321,11 +2321,8 @@
 CREATE FOREIGN TABLE f_t1(a int, b int) 
 	server database_16219 options(schema_name 'public', table_name 't1');
 select count(*) from f_t1;
- count 
--------
-    10
-(1 row)
-
+DETAIL:  FATAL:  no pg_hba.conf entry for host "::1", user "u16219", database "database_16219", SSL off
+ERROR:  could not connect to server "database_16219"
 DO $d$
     BEGIN
         EXECUTE $$ALTER SERVER database_16219
@@ -2333,11 +2330,8 @@
     END;
 $d$;
 select count(*) from f_t1;
- count 
--------
-    10
-(1 row)
-
+DETAIL:  FATAL:  no pg_hba.conf entry for host "::1", user "u16219", database "database_16219", SSL off
+ERROR:  could not connect to server "database_16219"
 \c postgres
 drop database database_16219;
 drop user u16219;

++ exit 1
```

- Fix the symbols' visibility and the linking process for MacOS.

+ @kainwen Since we're changing the visibility of symbols for the postgres binary and breaking ABIs and it's not a small change, any comments on this PR?

lmzzzzz1 commented:
> Hi, @lmzzzzz1 I added an additional commit to fix the failure mentioned above. Could you please take a look? I'm not a Makefile expert so I may make mistakes :-P
> 
> Symbols like `pqsignal` and those whose name starts with `pg_` should be exported both in `libpq.so` and the postgres binary. Things left us to do:
> 
> * Can we simplify the test? The current test case is failing on our CI system.
> 
> ```diff
> ======================================================================
> DIFF FILE: ./contrib/postgres_fdw/regression.diffs
> ----------------------------------------------------------------------
> 
> diff -I HINT: -I CONTEXT: -I GP_IGNORE: -U3 /home/gpadmin/gpdb_src/contrib/postgres_fdw/expected/gp_postgres_fdw_optimizer.out /home/gpadmin/gpdb_src/contrib/postgres_fdw/results/gp_postgres_fdw.out
> --- /home/gpadmin/gpdb_src/contrib/postgres_fdw/expected/gp_postgres_fdw_optimizer.out	2023-08-23 05:22:54.418706466 +0000
> +++ /home/gpadmin/gpdb_src/contrib/postgres_fdw/results/gp_postgres_fdw.out	2023-08-23 05:22:54.482712330 +0000
> @@ -2321,11 +2321,8 @@
>  CREATE FOREIGN TABLE f_t1(a int, b int) 
>  	server database_16219 options(schema_name 'public', table_name 't1');
>  select count(*) from f_t1;
> - count 
> --------
> -    10
> -(1 row)
> -
> +DETAIL:  FATAL:  no pg_hba.conf entry for host "::1", user "u16219", database "database_16219", SSL off
> +ERROR:  could not connect to server "database_16219"
>  DO $d$
>      BEGIN
>          EXECUTE $$ALTER SERVER database_16219
> @@ -2333,11 +2330,8 @@
>      END;
>  $d$;
>  select count(*) from f_t1;
> - count 
> --------
> -    10
> -(1 row)
> -
> +DETAIL:  FATAL:  no pg_hba.conf entry for host "::1", user "u16219", database "database_16219", SSL off
> +ERROR:  could not connect to server "database_16219"
>  \c postgres
>  drop database database_16219;
>  drop user u16219;
> 
> ++ exit 1
> ```
> 
> * Fix the symbols' visibility and the linking process for MacOS.
> 
> + @kainwen Since we're changing the visibility of symbols for the postgres binary and breaking ABIs and it's not a small change, any comments on this PR?

The modification to the Makefile looks good to me. 
- The CI failure seems to be caused by `localhost` being resolved to `::1` of ipv6, so maybe adding a line of rules to the `pg_hba.conf` file can solve it?
- I don't have a MacOS development environment locally, so I don't know if this BUG also exists on MacOS

zhaorui20005 commented:
For mac os, I compiled with such an error:
```
ld: unknown options: --version-script=postgres_symbols.map
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
Since there are no similar option for `-version-script` in mac, I still think the we had better fix this by adding 
```
if (pg_link_canary_is_frontend())
```
Then we made another fix for all platform in the future.

higuoxing commented:
> For mac os, I compiled with such an error:
> 
> ```
> ld: unknown options: --version-script=postgres_symbols.map
> clang: error: linker command failed with exit code 1 (use -v to see invocation)
> ```

Yeah, MacOS is using different linker with Linux, we might utilize the following options to control the symbol's visibility.

https://keith.github.io/xcode-man-pages/ld.1.html#Options_that_control_symbol_resolution

I'll give it a try.

> Since there are no similar option for `-version-script` in mac, I still think the we had better fix this by adding
> 
> ```
> if (pg_link_canary_is_frontend())
> ```
> 
> Then we make another fix for all platform in the future.

Sure, if we cannot finish it on time, it's the alternative solution.

zhaorui20005 commented on src/test/regress/GNUmakefile:
```diff
@@ -93,6 +93,9 @@ OBJS = $(WIN32RES) regress.o
 
 OBJS += regress_gp.o
 
+# TODO: comment.
```
Any new comments to be updated?

zhaorui20005 reviewed (approved):
The latest version of code look good to me. I have tested with my MAC os.
Thanks for the contribution.

lmzzzzz1 commented on src/backend/Makefile:
```diff
@@ -312,6 +338,9 @@ endif
 	rm -f postgres.dll libpostgres.a $(WIN32RES)
 endif
+ifeq (,$(filter cygwin win32 aix,$(PORTNAME)))
+	rm -f postgres_symbols.map
```
If darwin, hide_symbols.list should also be removed in make clean?

adam8157 commented:
So it uses the backend libpq while linking the binary postgres, but all the dynamically linked extensions which has SHLIB_LINK_INTERNAL = $(libpq) will match the frontend libpq symbols, am I understanding right?

higuoxing commented:
> So it uses the backend libpq while linking the binary postgres

No, some of the frontend codes are still compiled into the binary postgres due to Greenplum's MPP architecture (it needs to connect databases in backend processes). We haven't encountered problems described in #16219  in backend processes  so we just hide these symbols in this PR. Probably, we can teach the postgres binary to dynamically link against libpq.so or making Greenplum into an extension some day??

> but all the dynamically linked extensions which has `SHLIB_LINK_INTERNAL = $(libpq)` will match the frontend libpq symbols, am I understanding right?

Exactly.

adam8157 commented:
> > So it uses the backend libpq while linking the binary postgres
> 
> No, some of the frontend codes are still compiled into the binary postgres due to Greenplum's MPP architecture (it needs to connect databases in backend processes). We haven't encountered problems described in #16219 in backend processes so we just hide these symbols in this PR. 

Correct, it is a no as in there are some frontend files I soft-linked in PR https://github.com/greenplum-db/gpdb/pull/3608.

> Probably, we can teach the postgres binary to dynamically link against libpq.so or making Greenplum into an extension some day??

Since these frontend codes are almost the same (soft linked), I think it's likely safe. The difference is that it's only in binary or in the .so files.

> > but all the dynamically linked extensions which has `SHLIB_LINK_INTERNAL = $(libpq)` will match the frontend libpq symbols, am I understanding right?
> 
> Exactly.

adam8157 reviewed (approved):
Thanks for the contribution and collaboration.

kainwen commented:
This commit introduces a warning in my local env:

```
gawk: cmd. line:1: warning: regexp escape sequence `\#' is not a known regexp operator

gpadmin@gpdev:~/workspace/gpdb$ uname -a
Linux gpdev 5.15.0-79-generic #86-Ubuntu SMP Mon Jul 10 16:11:29 UTC 2023 aarch64 aarch64 aarch64 GNU/Linux
gpadmin@gpdev:~/workspace/gpdb$ cat /etc/issue
Ubuntu 22.04.2 LTS \n \l
```

Not look into. If you have time, please take a look. @higuoxing @lmzzzzz1

higuoxing commented:
> This commit introduces a warning in my local env:
> 
> ```
> gawk: cmd. line:1: warning: regexp escape sequence `\#' is not a known regexp operator
> 
> gpadmin@gpdev:~/workspace/gpdb$ uname -a
> Linux gpdev 5.15.0-79-generic #86-Ubuntu SMP Mon Jul 10 16:11:29 UTC 2023 aarch64 aarch64 aarch64 GNU/Linux
> gpadmin@gpdev:~/workspace/gpdb$ cat /etc/issue
> Ubuntu 22.04.2 LTS \n \l
> ```
> 
> Not look into. If you have time, please take a look. @higuoxing @lmzzzzz1

Thanks, it seems not a big issue, I'm going to look into it.
